File: python/pilotr/examples.py
"""Locate the design specifications shipped with pilotr.

The JSON files under ``pilotr/examples/`` are packaged copies of the
repository's ``spec/examples/``, shipped inside the wheel so they are reachable
from an installed copy. They are the same files that drive the R twin and the
no-code app, so a design authored once runs unchanged across all three.
"""

from __future__ import annotations

from importlib import resources
from typing import Optional, Union


def _examples_dir():
    return resources.files("pilotr") / "examples"


def _available() -> list[str]:
    return sorted(
        entry.name[:-5]
        for entry in _examples_dir().iterdir()
        if entry.name.endswith(".json")
    )


def pilotr_example(name: Optional[str] = None) -> Union[list[str], str]:
    """List the bundled example specifications, or return the path to one.

    pilotr ships one ready-to-run specification per design family, as JSON. These
    are the same files that drive the R twin and the no-code app.

    Parameters
    ----------
    name : str, optional
        The base name of an example, with or without the ``.json`` extension, for
        example ``"between_2group_gaussian"``. When ``None`` (the default), the
        available example names are returned instead of a path.

    Returns
    -------
    list of str, or str
        When ``name`` is ``None``, the available example names. Otherwise, the
        path to that example's JSON file, ready to pass to
        :func:`pilotr.simulate.load_spec`.
    """
    available = _available()
    if name is None:
        return available
    base = name[:-5] if name.endswith(".json") else name
    if base not in available:
        raise ValueError(
            f"Unknown example {name!r}. Available: {', '.join(available)}."
        )
    return str(_examples_dir() / f"{base}.json")

File: r/pilotr/R/examples.R
# Locate the bundled example specifications. The JSON files under inst/examples/
# are copies of the repository's spec/examples/, shipped inside the package so
# they are reachable with system.file() from an installed copy. They are the
# same files that drive the Python twin and the no-code app, so a design authored
# once runs unchanged across all three.

#' Example design specifications shipped with pilotr
#'
#' pilotr ships one ready-to-run specification per design family, as JSON, in the
#' package's `examples/` directory. These are the same files that drive the
#' Python twin and the no-code app, so a design authored in one place runs
#' unchanged in the others. `pilotr_example()` lists them, or returns the path to
#' one for [load_spec()].
#'
#' @param name The base name of an example, with or without the `.json`
#'   extension, for example `"between_2group_gaussian"`. When `NULL` (the
#'   default), the available example names are returned instead of a path.
#' @return When `name` is `NULL`, a character vector of the available example
#'   names. Otherwise, the full path to that example's JSON file, ready to pass
#'   to [load_spec()].
#' @seealso [load_spec()] to read a specification and
#'   [simulate_design()] to simulate from it.
#' @examples
#' pilotr_example()                       # the available examples
#' spec <- load_spec(pilotr_example("between_2group_gaussian"))
#' head(simulate_design(spec))
#' @export
pilotr_example <- function(name = NULL) {
  dir <- system.file("examples", package = "pilotr")
  available <- sort(tools::file_path_sans_ext(
    list.files(dir, pattern = "[.]json$")
  ))
  if (is.null(name)) {
    return(available)
  }
  if (!is.character(name) || length(name) != 1L || is.na(name) || !nzchar(name)) {
    stop("`name` must be a single example name, or NULL to list them.",
         call. = FALSE)
  }
  base <- tools::file_path_sans_ext(name)
  if (!base %in% available) {
    stop(sprintf("Unknown example '%s'. Available: %s.",
                 name, paste(available, collapse = ", ")), call. = FALSE)
  }
  system.file("examples", paste0(base, ".json"), package = "pilotr")
}
